core/session.ts

```diff
@@ -210,10 +216,19 @@ export class Session {
     });
   }
 
-  public resolve(name: string): string {
-    const table = this.tables[name];
+  public resolve(reference: string | IResolvable): string {
+    let [fQName, err] = [null, null];
+    if (typeof reference === "string") {
+      [fQName, err] = utils.matchFQName(reference, this.getAllFQNames());
```

Comment:
i think we can simplify this a lot, especially if we also do my suggestion above (to always set the `target` field):

when trying to `resolve` a `reference`, iterate through all actions (tables, operations, assertions) looking for actions whose `target.name` matches `reference` (if it's a string) or whose `target` matches `reference` (if it's an `IResolvable`). collect all matching actions. if matching action count is >1, throw a compile error complaining that the `reference` is not specific enough.

we can define all of this as a helper method on `Session` and then callers such as `Table.dependencies` can reuse it.

Reply:
1. This helper method which you mention is I _believe_ what I have called `utils.matchFQName`.  Unfortunately, it is also needed by the `validate` method to check for circular dependencies and ambiguous dependencies on validation time.  Since our long-term goal is to get rid of the validation process altogether and isolate core and api packages (see Issues https://github.com/dataform-co/dataform/issues/247, https://github.com/dataform-co/dataform/issues/107), I'd suggest we keep it as it is and once those two issues are addressed we can hopefully move `matchFQName`  from `utils` into `Session`. (I believe we _might_ be able to totally get rid of `utils` but that's another story).

2. But yes, I think `matchFQName` can now be largely simplified by using `{schema: string, name: string}` as a parameter, because now we can assume there will always be a target. I'll do that!